# Parcours Data Scientist - Yann Pham-Van

## Projet 3 : Concevez une application au service de la santé publique

**Mission**

L'agence "Santé publique France" a lancé **un appel à projets pour trouver des idées innovantes d’applications en lien avec l'alimentation**. Vous souhaitez y participer et proposer une idée d’application.

**Les données**

Extrait de l’appel à projets :

Le jeu de données Open Food Facts est disponible sur le site officiel (ou disponible à ce lien en téléchargement). Les variables sont définies à cette adresse.

Les champs sont séparés en quatre sections :

- Les informations générales sur la fiche du produit : nom, date de modification, etc.
- Un ensemble de tags : catégorie du produit, localisation, origine, etc.
- Les ingrédients composant les produits et leurs additifs éventuels.
- Des informations nutritionnelles : quantité en grammes d’un nutriment pour 100 grammes du produit.

**Votre mission**

Après avoir lu l’appel à projets, voici les différentes étapes que vous avez identifiées :

1) **Traiter le jeu de données**, en :

- **Réfléchissant à une idée d’application**.
- **Repérant des variables pertinentes** pour les traitements à venir, et nécessaires pour votre idée d’application.
- **Nettoyant les données** en :
    - mettant en évidence les éventuelles **valeurs manquantes**, avec au moins 3 méthodes de traitement adaptées aux variables concernées,
    - identifiant et en quantifiant les éventuelles **valeurs aberrantes** de chaque variable.
- **Automatisant ces traitements** pour éviter de répéter ces opérations

Le programme doit fonctionner si la base de données est légèrement modifiée (ajout d’entrées, par exemple).

2) Tout au long de l’analyse, **produire des visualisations** afin de mieux comprendre les données.

**Effectuer une analyse univariée** pour chaque variable intéressante, afin de synthétiser son comportement.

L’appel à projets spécifie que l’analyse doit être simple à comprendre pour un public néophyte. **Soyez donc attentif à la lisibilité** : taille des textes, choix des couleurs, netteté suffisante, et variez les graphiques (boxplots, histogrammes, diagrammes circulaires, nuages de points…) pour illustrer au mieux votre propos.

3) **Confirmer ou infirmer les hypothèses à l’aide d’une analyse multivariée. Effectuer les tests statistiques appropriés** pour vérifier la significativité des résultats.

4) **Justifier votre idée d’application**. Identifier des arguments justifiant la faisabilité (ou non) de l’application à partir des données Open Food Facts.

5) **Rédiger un rapport d’exploration** et **pitcher votre idée** durant la soutenance du projet.

# Nettoyage des données

Appel des librairies utilisées

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

## Inspection des données

Je crée le dataframe à partir du jeu de données.

In [2]:
df = pd.read_csv('fr.openfoodfacts.org.products.csv', sep = '\t', low_memory=False)

J'affiche les premières lignes des dataframes pour vérifier que les données sont bien chargées, puis leur structure et enfin le type des colonnes, la présence de doublons ou le taux de données manquantes.

In [3]:
df.head(5)

code                                                url  \
0  0000000003087  http://world-fr.openfoodfacts.org/produit/0000...   
1  0000000004530  http://world-fr.openfoodfacts.org/produit/0000...   
2  0000000004559  http://world-fr.openfoodfacts.org/produit/0000...   
3  0000000016087  http://world-fr.openfoodfacts.org/produit/0000...   
4  0000000016094  http://world-fr.openfoodfacts.org/produit/0000...   

                      creator   created_t      created_datetime  \
0  openfoodfacts-contributors  1474103866  2016-09-17T09:17:46Z   
1             usda-ndb-import  1489069957  2017-03-09T14:32:37Z   
2             usda-ndb-import  1489069957  2017-03-09T14:32:37Z   
3             usda-ndb-import  1489055731  2017-03-09T10:35:31Z   
4             usda-ndb-import  1489055653  2017-03-09T10:34:13Z   

  last_modified_t last_modified_datetime                    product_name  \
0      1474103893   2016-09-17T09:18:13Z              Farine de blé noir   
1      1489069957   2017-03-09T14:32:37Z  Banana Chips Sweetened (Whole)   
2      1489069957   2017-03-09T14:32:37Z                         Peanuts   
3      1489055731   2017-03-09T10:35:31Z          Organic Salted Nut Mix   
4      1489055653   2017-03-09T10:34:13Z                 Organic Polenta   

  generic_name quantity  ... ph_100g fruits-vegetables-nuts_100g  \
0          NaN      1kg  ...     NaN                         NaN   
1          NaN      NaN  ...     NaN                         NaN   
2          NaN      NaN  ...     NaN                         NaN   
3          NaN      NaN  ...     NaN                         NaN   
4          NaN      NaN  ...     NaN                         NaN   

  collagen-meat-protein-ratio_100g cocoa_100g chlorophyl_100g  \
0                              NaN        NaN             NaN   
1                              NaN        NaN             NaN   
2                              NaN        NaN             NaN   
3                              NaN        NaN             NaN   
4                              NaN        NaN             NaN   

  carbon-footprint_100g nutrition-score-fr_100g nutrition-score-uk_100g  \
0                   NaN                     NaN                     NaN   
1                   NaN                    14.0                    14.0   
2                   NaN                     0.0                     0.0   
3                   NaN                    12.0                    12.0   
4                   NaN                     NaN                     NaN   

  glycemic-index_100g water-hardness_100g  
0                 NaN                 NaN  
1                 NaN                 NaN  
2                 NaN                 NaN  
3                 NaN                 NaN  
4                 NaN                 NaN  

[5 rows x 162 columns]

In [4]:
df.shape

(320772, 162)

In [5]:
df.dtypes

code                        object
url                         object
creator                     object
created_t                   object
created_datetime            object
                            ...   
carbon-footprint_100g      float64
nutrition-score-fr_100g    float64
nutrition-score-uk_100g    float64
glycemic-index_100g        float64
water-hardness_100g        float64
Length: 162, dtype: object

In [6]:
df.duplicated().sum()

0

In [7]:
df.isna().mean().sort_values(ascending=False)

water-hardness_100g                      1.000000
no_nutriments                            1.000000
ingredients_that_may_be_from_palm_oil    1.000000
nutrition_grade_uk                       1.000000
nervonic-acid_100g                       1.000000
                                           ...   
created_datetime                         0.000028
created_t                                0.000009
creator                                  0.000006
last_modified_datetime                   0.000000
last_modified_t                          0.000000
Length: 162, dtype: float64

## Recherche des variables

Il me faut identifier les variables permettant :
- une utilisation par l'application lors de la phase d'exploration
- de catégoriser les individus/produits
- un taux de remplissage suffisant.

J'observe l'ensemble des variables à disposition.

In [8]:
colonnes = df.columns.tolist()
colonnes

['code',
 'url',
 'creator',
 'created_t',
 'created_datetime',
 'last_modified_t',
 'last_modified_datetime',
 'product_name',
 'generic_name',
 'quantity',
 'packaging',
 'packaging_tags',
 'brands',
 'brands_tags',
 'categories',
 'categories_tags',
 'categories_fr',
 'origins',
 'origins_tags',
 'manufacturing_places',
 'manufacturing_places_tags',
 'labels',
 'labels_tags',
 'labels_fr',
 'emb_codes',
 'emb_codes_tags',
 'first_packaging_code_geo',
 'cities',
 'cities_tags',
 'purchase_places',
 'stores',
 'countries',
 'countries_tags',
 'countries_fr',
 'ingredients_text',
 'allergens',
 'allergens_fr',
 'traces',
 'traces_tags',
 'traces_fr',
 'serving_size',
 'no_nutriments',
 'additives_n',
 'additives',
 'additives_tags',
 'additives_fr',
 'ingredients_from_palm_oil_n',
 'ingredients_from_palm_oil',
 'ingredients_from_palm_oil_tags',
 'ingredients_that_may_be_from_palm_oil_n',
 'ingredients_that_may_be_from_palm_oil',
 'ingredients_that_may_be_from_palm_oil_tags',
 'nutritio

La méthode pour afficher les taux de manquants, utilisée ci-dessus, ne permet pas une vision complète.

Pour trier les variables à retenir, il me faudrait une fonction pour afficher l'intégralité des variables et le taux de remplissage.

Elle sera réutilisée au fur et à mesure du nettoyage des données.

In [9]:
def calcul_remplissage(dataframe):
    
    dictionnaire = {}

    for colonne in dataframe.columns:
        dictionnaire[colonne] = []
        dictionnaire[colonne].append(round((1 - dataframe[colonne].isna().mean())*100, 1))
        dictionnaire[colonne].append(dataframe[colonne].isna().sum())

    return pd.DataFrame.from_dict(data=dictionnaire, orient='index', columns = ['Taux de remplissage', 'Nombre de manquants']).sort_values(by='Taux de remplissage', ascending=False)

In [10]:
pd.set_option('display.max_rows', 300)
calcul_remplissage(df)

Taux de remplissage  \
code                                                      100.0   
creator                                                   100.0   
created_t                                                 100.0   
created_datetime                                          100.0   
last_modified_t                                           100.0   
last_modified_datetime                                    100.0   
states_fr                                                 100.0   
states_tags                                               100.0   
states                                                    100.0   
url                                                       100.0   
countries                                                  99.9   
countries_fr                                               99.9   
countries_tags                                             99.9   
product_name                                               94.5   
brands_tags                                                91.1   
brands                                                     91.1   
energy_100g                                                81.4   
proteins_100g                                              81.0   
salt_100g                                                  79.7   
sodium_100g                                                79.6   
ingredients_that_may_be_from_palm_oil_n                    77.6   
ingredients_from_palm_oil_n                                77.6   
additives                                                  77.6   
additives_n                                                77.6   
ingredients_text                                           77.6   
sugars_100g                                                76.4   
fat_100g                                                   76.0   
carbohydrates_100g                                         75.9   
saturated-fat_100g                                         71.6   
nutrition_grade_fr                                         69.0   
nutrition-score-fr_100g                                    69.0   
nutrition-score-uk_100g                                    69.0   
serving_size                                               65.9   
fiber_100g                                                 62.6   
additives_tags                                             48.2   
additives_fr                                               48.2   
cholesterol_100g                                           44.9   
trans-fat_100g                                             44.7   
calcium_100g                                               44.0   
vitamin-c_100g                                             43.9   
iron_100g                                                  43.8   
vitamin-a_100g                                             42.9   
quantity                                                   32.7   
pnns_groups_2                                              29.5   
pnns_groups_1                                              28.5   
main_category_fr                                           26.3   
main_category                                              26.3   
categories_tags                                            26.3   
categories_fr                                              26.3   
categories                                                 26.3   
packaging_tags                                             24.6   
packaging                                                  24.6   
image_url                                                  23.6   
image_small_url                                            23.6   
purchase_places                                            18.1   
generic_name                                               16.5   
stores                                                     16.1   
labels_tags                                                14.5   
labels                                                     14.5   
labels_fr               

Le niveau de remplissage est très fluctuant. Peu de variables sont bien renseignées.
A peine 34 sur les 162 sont remplies à plus de 50% et tout juste un tiers (54) est carrément à 0% !!!

### **Idée d'application**

Au vu de ces résultats, une idée d'application commence à germer.
Il serait intéressant de se concentrer sur la partie négative du nutriscore afin de la minimiser en effectuant une cotation des composantes *énergie*, *graisses saturées*, *sucres* et *sodium*.

L'idée de l'application consisterait à déterminer la catégorie d'un produit alimentaire afin de suggérer les meilleures alternatives dans cette catégorie ou catégorie élargie, c'est-à-dire les moins caloriques, grasses, sucrées et salées.

**Sélection simple des variables**

Pour démarrer le nettoyage, plusieurs variables sont disponibles.

Je sélectionne celles qui vont me permettre d'identifier les produits et de les catégoriser :
- code : code-barres du produit
- product_name : nom du produit
- brands : marque du produit
- countries_fr : pays de vente du produit
- pnns_groups_1 et pnns_groups_2 : familles et sous-familles de produits définies dans le cadre du *Plan National Nutrition Santé* (PNNS)
- nutrition_grade_fr : classes du Nutri-Score

En réalité, il me faut absolument les variables de familles et sous-familles pour catégoriser un produit. Cela est nécessaire pour que l'application puisse suggérer d'autres produits similaires mieux classés.

In [11]:
df_categorisable = df.loc[df['product_name'].notna() & df['pnns_groups_1'].notna() & df['pnns_groups_2'].notna()]

print(f'Il subsiste {df_categorisable.shape[0]} sur {df.shape[0]} observations initiales.')
print(f'Cela représente un taux de {df_categorisable.shape[0]/df.shape[0]*100:.1f} %.')

Il subsiste 88498 sur 320772 observations initiales.
Cela représente un taux de 27.6 %.


Quelles sont les conséquences de ce premier filtre sur les taux de remplissage ?

In [12]:
calcul_remplissage(df_categorisable)

Taux de remplissage  \
code                                                      100.0   
product_name                                              100.0   
states_tags                                               100.0   
states                                                    100.0   
pnns_groups_2                                             100.0   
pnns_groups_1                                             100.0   
url                                                       100.0   
states_fr                                                 100.0   
last_modified_datetime                                    100.0   
creator                                                   100.0   
last_modified_t                                           100.0   
created_datetime                                          100.0   
created_t                                                 100.0   
countries                                                  99.9   
countries_tags                                             99.9   
countries_fr                                               99.9   
brands_tags                                                97.4   
brands                                                     97.4   
quantity                                                   91.1   
categories_fr                                              91.0   
main_category                                              91.0   
categories_tags                                            91.0   
categories                                                 91.0   
main_category_fr                                           91.0   
packaging_tags                                             82.8   
packaging                                                  82.8   
additives_n                                                79.1   
ingredients_text                                           79.1   
ingredients_that_may_be_from_palm_oil_n                    79.1   
additives                                                  79.1   
ingredients_from_palm_oil_n                                79.1   
energy_100g                                                75.7   
proteins_100g                                              75.0   
fat_100g                                                   71.9   
carbohydrates_100g                                         71.4   
salt_100g                                                  71.3   
sodium_100g                                                71.3   
sugars_100g                                                71.1   
saturated-fat_100g                                         70.7   
image_url                                                  69.8   
image_small_url                                            69.8   
nutrition-score-uk_100g                                    69.0   
nutrition_grade_fr                                         69.0   
nutrition-score-fr_100g                                    69.0   
purchase_places                                            62.9   
generic_name                                               57.8   
stores                                                     56.2   
labels_fr                                                  47.3   
labels_tags                                                47.3   
labels                                                     47.2   
fiber_100g                                                 46.1   
additives_fr                                               43.3   
additives_tags                                             43.3   
manufacturing_places_tags                                  39.6   
manufacturing_places                                       39.6   
serving_size                                               39.1   
emb_codes_tags                                             32.4   
emb_codes                                                  32.4   
allergens                                                  30.8   
traces                  

Maintenant que tous les produits restants peuvent être catégoriser, je m'intéresse aux variables supplémentaires utilisables par l'application.

Aux 7 précédemment identifiées, je retiens les données quantitatives relatives directement ou indirectement au Nutri-Score, soit les  variables supplémentaires :
- energy_100g
- saturated-fat_100g
- sodium_100g
- sugars_100g
- proteins_100g
- fiber_100g
- fruits-vegetables-nuts_100g
- fat_100g
- nutrition-score-fr_100g

In [13]:
variables_quantitatives = ['energy_100g', 'saturated-fat_100g', 'sodium_100g', 'sugars_100g', 'proteins_100g', 'fiber_100g',
             'fruits-vegetables-nuts_100g', 'fat_100g', 'nutrition-score-fr_100g']
variables = ['code', 'product_name', 'brands', 'countries_fr', 'pnns_groups_1', 'pnns_groups_2', 'nutrition_grade_fr'] + variables_quantitatives
df_filtre = df_categorisable[variables]
df_filtre = df_filtre.reset_index(drop=True)

In [14]:
df_filtre.head()

code    product_name              brands       countries_fr  \
0  0000000024600   Filet de bœuf                 NaN             France   
1  0000000036252  Lion Peanut x2            Sunridge  France,États-Unis   
2  0000000039259         Twix x2                 NaN             France   
3  0000000039529  Pack de 2 Twix      Twix, Lundberg  France,États-Unis   
4  0000000290616    Salade Cesar  Kirkland Signature             Canada   

           pnns_groups_1 pnns_groups_2 nutrition_grade_fr  energy_100g  \
0                unknown       unknown                NaN          NaN   
1                unknown       unknown                  e       1883.0   
2                unknown       unknown                NaN          NaN   
3                unknown       unknown                NaN       1481.0   
4  Fruits and vegetables    Vegetables                  c       1210.0   

   saturated-fat_100g  sodium_100g  sugars_100g  proteins_100g  fiber_100g  \
0                 NaN          NaN          NaN            NaN         NaN   
1                12.5        0.038         57.5           2.50         2.5   
2                 NaN          NaN          NaN            NaN         NaN   
3                 NaN          NaN          NaN           6.25         6.2   
4                 7.0        0.850          0.0          22.00         2.0   

   fruits-vegetables-nuts_100g  fat_100g  nutrition-score-fr_100g  
0                          NaN       NaN                      NaN  
1                          NaN     20.00                     22.0  
2                          NaN       NaN                      NaN  
3                          NaN      4.17                      NaN  
4                          NaN     12.00                      6.0

In [15]:
calcul_remplissage(df_filtre)

Taux de remplissage  Nombre de manquants
code                                       100.0                    0
product_name                               100.0                    0
pnns_groups_1                              100.0                    0
pnns_groups_2                              100.0                    0
countries_fr                                99.9                  131
brands                                      97.4                 2268
energy_100g                                 75.7                21461
proteins_100g                               75.0                22116
fat_100g                                    71.9                24863
sodium_100g                                 71.3                25365
sugars_100g                                 71.1                25620
saturated-fat_100g                          70.7                25912
nutrition_grade_fr                          69.0                27391
nutrition-score-fr_100g                     69.0                27391
fiber_100g                                  46.1                47658
fruits-vegetables-nuts_100g                  3.4                85471

Je note que même si les familles et sous-familles sont remplies à 100%, il apparaît des valeurs *unknown*.

**Traitement des *pnns_groups_1* et *pnns_groups_2***

Il s'agit des valeurs *unknown* qu'il me faut quantifier.

In [16]:
pnns1_unknown = df_filtre.loc[df_filtre['pnns_groups_1'] == 'unknown'].shape[0]
print('Il y a', pnns1_unknown, 'observations avec unknown comme valeur de pnns_groups_1')
pnns2_unknown = df_filtre.loc[df_filtre['pnns_groups_2'] == 'unknown'].shape[0]
print('Il y a', pnns2_unknown, 'observations avec unknown comme valeur de pnns_groups_2')

Il y a 20154 observations avec unknown comme valeur de pnns_groups_1
Il y a 20154 observations avec unknown comme valeur de pnns_groups_2


Je supprime toutes les observations présentes ou futures concernées.

In [17]:
df_filtre = df_filtre.loc[(df_filtre['pnns_groups_1'] != 'unknown') & (df_filtre['pnns_groups_2'] != 'unknown')]
print(f'Il subsiste {df_filtre.shape[0]} sur {df.shape[0]} observations initiales.')
print(f'Cela représente un taux de {df_filtre.shape[0]/df.shape[0]*100:.1f} %.')

Il subsiste 68344 sur 320772 observations initiales.
Cela représente un taux de 21.3 %.


Voyons les modalités prises :

In [18]:
df_filtre['pnns_groups_1'].unique()

array(['Fruits and vegetables', 'Sugary snacks', 'Cereals and potatoes',
       'Composite foods', 'Fish Meat Eggs', 'Beverages', 'Fat and sauces',
       'fruits-and-vegetables', 'Milk and dairy products', 'Salty snacks',
       'sugary-snacks', 'cereals-and-potatoes', 'salty-snacks'],
      dtype=object)

In [19]:
df_filtre['pnns_groups_2'].unique()

array(['Vegetables', 'Biscuits and cakes', 'Bread', 'Legumes',
       'Pizza pies and quiche', 'Meat', 'Non-sugared beverages', 'Sweets',
       'Sweetened beverages', 'Dressings and sauces', 'One-dish meals',
       'vegetables', 'Soups', 'Chocolate products', 'Fruits', 'Cereals',
       'Milk and yogurt', 'Fats', 'Cheese', 'Sandwich', 'Appetizers',
       'Nuts', 'Breakfast cereals', 'Artificially sweetened beverages',
       'Fruit juices', 'Eggs', 'Fish and seafood', 'Dried fruits',
       'Ice cream', 'Processed meat', 'Potatoes', 'Dairy desserts',
       'Fruit nectars', 'pastries', 'fruits', 'Salty and fatty products',
       'cereals', 'legumes', 'nuts'], dtype=object)

Je constate des variations sur des modalités similaires dans ces variables.
Par exemple :
- sugary-snacks
- Sugary snacks

Des modalités ont une variante avec tiret et tout en minuscules.

Je transforme les variantes en modalités d'origine.

In [20]:
df_filtre['pnns_groups_1'] = df_filtre['pnns_groups_1'].str.capitalize().str.replace('-', ' ')
df_filtre['pnns_groups_2'] = df_filtre['pnns_groups_2'].str.capitalize().str.replace('-', ' ')

J'ai aussi un doute avec la modalité *Legumes* qui pourrait être une traduction en français pour *Vegetables* mais pourrait aussi désigner, en anglais, la famille des *Légumineuses*, je vérifie d'abord.

In [21]:
df_filtre.loc[df_filtre['pnns_groups_2'] == 'Legumes'].sample(10)

code                           product_name  \
57743  3760042970040                    haricots mange tout   
62115  4010355208576                             Erdnussmus   
41434  3351700067112                  Peanut Butter Crunchy   
82038  8422584000351                    Tofu estilo japonés   
62857  4019738103491                    Räuchertofu Premium   
65772  4388844021389                   Erdnussbutter creamy   
83887  8480017012333    Alubias blancas cocidas en conserva   
21682  3222474196547  Lentilles corail Bio - 500 g - Casino   
38965  3307130000199            Haricots rouges biologiques   
34968  3268387010501                           Pois chiches   

                                                  brands  \
57743                                          Le Marché   
62115                                             dm Bio   
41434                                            Duerr s   
82038  Ecocesta Productos Ecológicos,//Propiedad de:/...   
62857                                        veggie life   
65772                                           REWE Bio   
83887  Dia,//Propiedad de://,Dia - Distribuidora Inte...   
21682                                             Casino   
38965                                           Bioviver   
34968                              Conserverie casatorra   

                                countries_fr         pnns_groups_1  \
57743                                 France  Cereals and potatoes   
62115  Autriche,République tchèque,Allemagne  Cereals and potatoes   
41434                                 France  Cereals and potatoes   
82038                                Espagne  Cereals and potatoes   
62857                              Allemagne  Cereals and potatoes   
65772                              Allemagne  Cereals and potatoes   
83887                          Maroc,Espagne  Cereals and potatoes   
21682                                 France  Cereals and potatoes   
38965                                 France  Cereals and potatoes   
34968                                 France  Cereals and potatoes   

      pnns_groups_2 nutrition_grade_fr  energy_100g  saturated-fat_100g  \
57743       Legumes                NaN          NaN                 NaN   
62115       Legumes                  c       2667.0                 7.7   
41434       Legumes                  d       2385.0                 8.0   
82038       Legumes                  a        615.0                 1.3   
62857       Legumes                  d        839.0                 1.9   
65772       Legumes                  e       2630.0                10.6   
83887       Legumes                NaN        336.0                 NaN   
21682       Legumes                  a       1320.0                 0.3   
38965       Legumes                  a        308.0                 0.1   
34968       Legumes                NaN          NaN                 NaN   

       sodium_100g  sugars_100g  proteins_100g  fiber_100g  \
57743          NaN          NaN            NaN         NaN   
62115     0.007874         4.50           25.0         8.6   
41434     0.255906        10.00           21.0         0.0   
82038     0.039370         0.03           15.2         1.6   
62857     0.787000         1.00           21.7         NaN   
65772     0.393701         7.30           26.6         NaN   
83887          NaN          NaN            4.8         NaN   
21682     0.010000         2.50           23.0        19.0   
38965     0.153543         0.40            5.3         4.8   
34968          NaN          NaN            NaN         NaN   

       fruits-vegetables-nuts_100g  fat_100g  nutrition-score-fr_100g  
57743                          NaN       NaN                      NaN  
62115                          NaN      50.0                      9.0  
41434                          NaN       NaN                     18.0  
82038                          NaN       8.5                     -5.0  
62857                

Il s'agit bien de légumineuses donc il faut laisser cette distinction.

Je place les familles et sous-familles dans des listes pour pourront m'être utiles plus tard.

In [22]:
familles = df_filtre['pnns_groups_1'].unique().tolist()
sous_familles = df_filtre['pnns_groups_2'].unique().tolist()

## Traitement des valeurs aberrantes

**Données nutritionnelles**

Par nature, les données nutritionnelles sont un pourcentage, sauf pour l'énergie fournie en kJ.

Les valeurs aberrantes sont celles négatives ou >100.
Je vérifie à quoi ressemblent ces éventuelles observations.

In [23]:
outliers_nutrition = df_filtre.loc[(df_filtre['saturated-fat_100g'] > 100) | (df_filtre['saturated-fat_100g'] < 0) 
              | (df_filtre['sodium_100g'] > 100) | (df_filtre['sodium_100g'] < 0) 
              | (df_filtre['sugars_100g'] > 100) | (df_filtre['sugars_100g'] < 0) 
              | (df_filtre['proteins_100g'] > 100) | (df_filtre['proteins_100g'] < 0) 
              | (df_filtre['fat_100g'] > 100) | (df_filtre['fat_100g'] < 0) 
              | (df_filtre['fiber_100g'] > 100) | (df_filtre['fiber_100g'] < 0) 
              | (df_filtre['fruits-vegetables-nuts_100g'] > 100) | (df_filtre['fruits-vegetables-nuts_100g'] < 0)]
outliers_nutrition

code                                product_name  \
2849        01364008                              Tomato Ketchup   
6026   2000000045416                                       flake   
17980  3161712000928                           Caprice des dieux   
49003  3560070740338                               Sirop d'agave   
53352  3596710288755  mini choux goût fromage de chèvre - poivre   
69702  5060224881163                               Cheese salad    
79181  8005305900255                    Ekstra Jomfru Olivenolie   
79979  8032942610032                    Graine de couscous moyen   

                  brands countries_fr            pnns_groups_1  \
2849               Heinz   États-Unis           Fat and sauces   
6026                 NaN  en:السعودية            Sugary snacks   
17980  Caprice des Dieux       France  Milk and dairy products   
49003      Carrefour Bio       France            Sugary snacks   
53352             Auchan       France             Salty snacks   
69702                NaN  Royaume-Uni          Composite foods   
79181          Santagata     Danemark           Fat and sauces   
79979    La méditerranéa       France     Cereals and potatoes   

              pnns_groups_2 nutrition_grade_fr  energy_100g  \
2849   Dressings and sauces                  d       2510.0   
6026     Chocolate products                  e        519.0   
17980                Cheese                  d       1379.0   
49003                Sweets                NaN       1785.0   
53352            Appetizers                  e      18700.0   
69702              Sandwich                  d        348.0   
79181                  Fats                  d       3737.0   
79979               Cereals                  a       1482.0   

       saturated-fat_100g  sodium_100g  sugars_100g  proteins_100g  \
2849                0.588     0.004170        134.0           7.06   
6026               17.700   109.000000         57.6           8.50   
17980              21.000     0.551181         -0.1          15.30   
49003                 NaN          NaN        105.0            NaN   
53352             210.000     3.670000         22.7           0.00   
69702               4.800   117.165354          0.4           3.60   
79181              15.000     0.000000          0.0           0.00   
79979               0.300     0.003937          2.5          12.00   

       fiber_100g  fruits-vegetables-nuts_100g  fat_100g  \
2849          0.0                          NaN     0.588   
6026          1.9                          NaN    28.500   
17980         NaN                          NaN    30.000   
49003         NaN                          NaN       NaN   
53352         0.0                          NaN   380.000   
69702         0.1                          NaN     4.700   
79181         NaN                          NaN   101.000   
79979         NaN                          NaN   105.000   

       nutrition-score-fr_100g  
2849                      17.0  
6026                      29.0  
17980                     15.0  
49003                      NaN  
53352                     35.0  
69702                     15.0  
79181                     11.0  
79979                     -1.0

Il y a finalement peu d'observations correspondantes.

Quelques unes pourraient être récupérées, comme le fromage *Caprice des dieux* avec -0.1% en sucres ou les sirops d'agave.

Mais vu le faible nombre et dans le cadre d'un **traitement automatisé**, je ne ferai pas de cas par cas.

In [24]:
df_filtre.drop(index=outliers_nutrition.index, inplace=True)
print(f'Il subsiste {df_filtre.shape[0]} sur {df.shape[0]} observations initiales.')
print(f'Cela représente un taux de {df_filtre.shape[0]/df.shape[0]*100:.1f} %.')

Il subsiste 68336 sur 320772 observations initiales.
Cela représente un taux de 21.3 %.


**Variable énergétique *energy_100g***

Diverses sources mentionnent les aliments les plus caloriques jusqu'à 900kcal/100g, soit 3765 kJ/100g.
Je fixe comme outliers les valeurs négatives ou au-delà de 3800.

In [25]:
outliers_energie = df_filtre.loc[(df_filtre['energy_100g'] > 3800) | (df_filtre['energy_100g'] < 0)]
outliers_energie

code                                       product_name  \
1468   0051500239131                    Crisco All-Vegetable Shortening   
6028   2000000045489                                             bubbly   
7308        20233679                                 Halva with Almonds   
9004        20842437                                      Lardons fumés   
10375       26009575                         Pruneaux d'Agen Dénoyautés   
16241  3092718605216                                    Sirop de cerise   
16627  3101740020017                  Confiture de Châtaignes d'Ardèche   
23623  3245390220660                   Rillettes de sardine de Bretagne   
24609  3250390001522                  Sirop Grenadine - Recette William   
26570  3250392034757                                  Pains Au Chocolat   
28578  3256221675573                                       Purée Nature   
30404  3257981937024                   Mini-bonbons Sans Sucres, Menthe   
31074  3258561300597                                    Sirop de Menthe   
35037  3270020000183                          Confiture pastèque citron   
38222  3291960006127  Huiles vierge biologique de Colza - première p...   
38923  3303640001404                                   Croûtons à l'ail   
45773  3473710080032                                  Palets pur beurre   
49560  3560070927296                 Emmental français râpé au lait cru   
54307  3596710396115            Sardines à l'huile d'olive vierge extra   
54308  3596710396122                    Sardines A L'huile De Tournesol   
54833  3596710440634                                         Mozzarella   
57008  3700640470139                Gruyère France au lait cru (33% MG)   
81693  8414606446261  Aceite de oliva virgen extra &quot;Esencia And...   
84215  8480017111852                       Chocolate negro sin azúcares   

                                                  brands countries_fr  \
1468                                              Crisco   États-Unis   
6028                                          dairy milk  en:السعودية   
7308                                      Eridanous,Lidl       France   
9004                                          Saint Alby       France   
10375                                       Golden Fruit       France   
16241                                          Teisseire       France   
16627                                            Sabaton       France   
23623                                  Reflets de France       France   
24609                                            Paquito       France   
26570                                              Netto       France   
28578                                            U bio,U       France   
30404                                               Cora       France   
31074                                       Belle France       France   
35037                               Vergers des Alpilles       France   
38222                                         Emile noel       France   
38923                                       Marie-Amélie       France   
45773                                      Georges Blanc       France   
49560                            Carrefour Bio,Carrefour       France   
54307                                             Auchan       France   
54308                                             Auchan       France   
54833                                             Auchan       France   
57008                           R Bio,Fromagerie Roussey       France   
81693  Esencia Andalusí,//Propiedad de://,José Serran...      Espagne   
84215                                     Vital Dia, Dia      Espagne   

                 pnns_groups_1          pnns_groups_2 nutrition_grade_fr  \
1468            Fat and sauces                   Fats                NaN   
6028             Sugary snacks     Chocolate products                  e   
7308             Sugary snacks                 Sweets                  e   
9004      

In [26]:
outliers_energie.shape

(24, 16)

Je supprime ces observations.

In [27]:
df_filtre.drop(index=outliers_energie.index, inplace=True)
print(f'Il subsiste {df_filtre.shape[0]} sur {df.shape[0]} observations initiales.')
print(f'Cela représente un taux de {df_filtre.shape[0]/df.shape[0]*100:.1f} %.')

Il subsiste 68312 sur 320772 observations initiales.
Cela représente un taux de 21.3 %.


**Score nutritionnel *nutrition-score-fr_100g***

Ce score est mathématiquement borné entre -15 et +40.
Quelles observations sont concernées ?

In [28]:
outliers_nutriscore = df_filtre.loc[(df_filtre['nutrition-score-fr_100g'] > 40) | (df_filtre['nutrition-score-fr_100g'] < -15)]
outliers_nutriscore

Empty DataFrame
Columns: [code, product_name, brands, countries_fr, pnns_groups_1, pnns_groups_2, nutrition_grade_fr, energy_100g, saturated-fat_100g, sodium_100g, sugars_100g, proteins_100g, fiber_100g, fruits-vegetables-nuts_100g, fat_100g, nutrition-score-fr_100g]
Index: []

Parfait, la variable est bien cadrée.
Dans le cadre d'une automatisation, je procède à la suppression d'éventuelles futures entrées erronées.

In [29]:
df_filtre.drop(index=outliers_nutriscore.index, inplace=True)
print(f'Il subsiste {df_filtre.shape[0]} sur {df.shape[0]} observations initiales.')
print(f'Cela représente un taux de {df_filtre.shape[0]/df.shape[0]*100:.1f} %.')

Il subsiste 68312 sur 320772 observations initiales.
Cela représente un taux de 21.3 %.


**Variables qualitatives**

Pour aborder les variables de nature qualitative, je continue sur les classes du nutriscore ***nutrition_grade_fr***.
Quelles sont les différentes modalités ?

In [30]:
classes_nutriscore = df_filtre['nutrition_grade_fr'].unique().tolist()
classes_nutriscore

['c', nan, 'b', 'e', 'd', 'a']

Parfait, là aussi : il n'y a que les classes de valeurs attendues ou des manquants qu'il me sera peut-être possible de recalculer.

Je veille à supprimer toute autre éventuelle future entrée non conforme.

In [31]:
df_filtre = df_filtre.loc[(df_filtre['nutrition_grade_fr'].isin(['e', 'c', 'b', 'd', 'a'])) | (df_filtre['nutrition_grade_fr'].isna())]

## Traitement des dernières valeurs manquantes

Je récapitule les taux de remplissage des variables.

In [32]:
calcul_remplissage(df_filtre)

Taux de remplissage  Nombre de manquants
code                                       100.0                    0
product_name                               100.0                    0
pnns_groups_1                              100.0                    0
pnns_groups_2                              100.0                    0
countries_fr                                99.9                   51
brands                                      98.8                  808
energy_100g                                 80.1                13584
proteins_100g                               79.4                14055
fat_100g                                    76.7                15912
sodium_100g                                 75.7                16588
sugars_100g                                 75.5                16739
saturated-fat_100g                          75.2                16940
nutrition_grade_fr                          73.7                17978
nutrition-score-fr_100g                     73.7                17978
fiber_100g                                  48.3                35299
fruits-vegetables-nuts_100g                  3.3                66076

**Variable *countries_fr***

La méconnaissance du pays de vente, surtout en si faible proportion, n'est pas un frein à la mise en oeuvre de l'application.
Je décide d'affecter la valeur *Inconnu* à ces observations.

In [33]:
df_filtre.loc[df_filtre['countries_fr'].isna(), 'countries_fr'] = 'Inconnu'

**Variable *brands***

Voyons de quoi il retourne :

In [34]:
df_filtre.loc[df_filtre['brands'].isna()]

code                         product_name brands countries_fr  \
44          00008761                        Spring onions    NaN  Royaume-Uni   
99     0007200000021                     Sauce bolognaise    NaN       Suisse   
116    0009138378043                       Rainbow Cherry    NaN   États-Unis   
135    0010700531001                            Milk Duds    NaN   États-Unis   
240    0013000001243                       Ketchup Heinze    NaN      Inconnu   
...              ...                                  ...    ...          ...   
88109  9315748110005         McLaren Vale Free Range Eggs    NaN    Australie   
88215  9332907005101  Sweet Baby Munch and Crunch Carrots    NaN    Australie   
88262  9342508000061                Strawberry cheesecake    NaN    Australie   
88271  9344378000059               Chicken tandoori pizza    NaN    Australie   
88295  9348215874407    Saucisson de cheval amylacé rouge    NaN       France   

               pnns_groups_1          pnns_groups_2 nutrition_grade_fr  \
44     Fruits and vegetables             Vegetables                NaN   
99            Fat and sauces   Dressings and sauces                NaN   
116            Sugary snacks                 Sweets                NaN   
135            Sugary snacks                 Sweets                  e   
240           Fat and sauces   Dressings and sauces                NaN   
...                      ...                    ...                ...   
88109         Fish meat eggs                   Eggs                  b   
88215  Fruits and vegetables             Vegetables                NaN   
88262          Sugary snacks     Biscuits and cakes                NaN   
88271        Composite foods  Pizza pies and quiche                NaN   
88295         Fish meat eggs         Processed meat                NaN   

       energy_100g  saturated-fat_100g  sodium_100g  sugars_100g  \
44             NaN                 NaN          NaN          NaN   
99             NaN                 NaN          NaN          NaN   
116            NaN                 NaN          NaN          NaN   
135         1820.0                8.97        0.179         51.3   
240            NaN                 NaN          NaN          NaN   
...            ...                 ...          ...          ...   
88109        559.0                3.30        0.136          0.3   
88215          NaN                 NaN          NaN          NaN   
88262          NaN                 NaN          NaN          NaN   
88271          NaN                 NaN          NaN          NaN   
88295          NaN                 NaN          NaN          NaN   

       proteins_100g  fiber_100g  fruits-vegetables-nuts_100g  fat_100g  \
44               NaN         NaN                          NaN       NaN   
99               NaN         NaN                          NaN       NaN   
116              NaN         NaN                          NaN       NaN   
135             2.56         0.0                          NaN      15.4   
240              NaN         NaN                          NaN       NaN   
...              ...         ...                          ...       ...   
88109          12.20         NaN                          NaN       9.9   
88215            NaN         NaN                          NaN       NaN   
88262            NaN         NaN                          NaN       NaN   
88271            NaN         NaN                          NaN       NaN   
88295            NaN         NaN                          NaN       NaN   

       nutrition-score-fr_100g  
44                         NaN  
99                         NaN  
116                        NaN  
135                       24.0  
240                        NaN  
...                        ...  
88109                      0.0  
88215                      NaN  
88262                      NaN  
88271                      NaN  
88295                      NaN  

[808 rows x 16 columns]

De même, la marque est moins primordiale que le nom du produit et il serait dommage de supprimer une observation pour ce manquant.

Je décide d'affecter le nom de produit à la marque, quand elle manque.

In [35]:
df_filtre.loc[df_filtre['brands'].isna(), 'brands'] = df_filtre['product_name']

Je laisse de côté la dernière variable qualitative contenant encore des manquants *nutrition_grade_fr* car elle dépend directement de la variable quantitative *nutrition-score-fr_100g* qu'il sera possible reconstruire.

**variables quantitatives**

Je souhaite implémenter la médiane des valeurs par sous-famille.
J'observe ces médianes :

In [36]:
df_filtre.groupby('pnns_groups_2')[variables_quantitatives].median()

energy_100g  saturated-fat_100g  \
pnns_groups_2                                                       
Appetizers                            2110.00              3.1000   
Artificially sweetened beverages        19.75              0.0000   
Biscuits and cakes                    1936.00              9.6000   
Bread                                 1240.50              0.6000   
Breakfast cereals                     1640.00              1.7000   
Cereals                               1489.00              0.4755   
Cheese                                1253.00             17.0000   
Chocolate products                    2273.00             19.0000   
Dairy desserts                         573.00              3.3000   
Dressings and sauces                   611.00              1.0000   
Dried fruits                          1305.00              0.2000   
Eggs                                   602.00              2.7000   
Fats                                  3077.00             16.0000   
Fish and seafood                       746.50              2.0000   
Fruit juices                           196.50              0.0000   
Fruit nectars                          200.00              0.0000   
Fruits                                 276.00              0.0500   
Ice cream                              899.00              7.0000   
Legumes                                579.00              0.2750   
Meat                                   656.00              1.9000   
Milk and yogurt                        350.00              1.9000   
Non sugared beverages                  160.00              0.0100   
Nuts                                  2561.50              6.4000   
One dish meals                         565.00              1.4000   
Pastries                              1512.00              7.4000   
Pizza pies and quiche                  976.00              4.4000   
Potatoes                               402.00              0.5000   
Processed meat                        1033.50              6.8000   
Salty and fatty products              1879.00              5.0000   
Sandwich                              1020.00              3.2000   
Soups                                  159.00              0.5000   
Sweetened beverages                    176.00              0.0000   
Sweets                                1498.00              0.2000   
Vegetables                             133.00              0.1000   

                                  sodium_100g  sugars_100g  proteins_100g  \
pnns_groups_2                                                               
Appetizers                           0.629921        2.400            6.4   
Artificially sweetened beverages     0.009090        0.785            0.0   
Biscuits and cakes                   0.224409       30.550            6.1   
Bread                                0.472441        4.200            9.2   
Breakfast cereals                    0.157480       21.900            8.5   
Cereals                              0.012598        2.400           10.5   
Cheese                               0.551181        0.500           18.0   
Chocolate products                   0.051181       47.000            6.7   
Dairy desserts                       0.051181       16.000            3.4   
Dressings and sauces                 0.630000        4.800            1.5   
Dried fruits                         0.011811       49.000            2.7   
Eggs                                 0.125984        0.700           12.6   
Fats                                 0.011811        0.000            0.0   
Fish and seafood                     0.472441        0.400           18.2   
Fruit juices                         0.003937       10.000            0.5   
Fruit nectars                        0.003937       10.700            0.3   
Fruits                               0.003937       13.300            0.5   
Ice cream                            0.050000       24.000            2.8   
Legumes  

Je ne note pas d'incohérence mais je remarque 2 manquants en *fruits-vegetables-nuts_100g* pour *Potatoes* et *Salty and fatty products*. Il faudra mettre ces valeurs à 0 plus tard.

En attendant, j'effectue le remplacement de tous les manquants restants par la médiane de sous-famille.

In [37]:
df_filtre.fillna(df_filtre.groupby('pnns_groups_2').transform('median'),inplace=True)

C:\Users\Yann\AppData\Local\Temp\ipykernel_3696\1088108492.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.median is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df_filtre.fillna(df_filtre.groupby('pnns_groups_2').transform('median'),inplace=True)


Les taux de remplacement ont dû bien évoluer.

In [38]:
calcul_remplissage(df_filtre)

Taux de remplissage  Nombre de manquants
code                                       100.0                    0
product_name                               100.0                    0
brands                                     100.0                    0
countries_fr                               100.0                    0
pnns_groups_1                              100.0                    0
pnns_groups_2                              100.0                    0
energy_100g                                100.0                    0
saturated-fat_100g                         100.0                    0
sodium_100g                                100.0                    0
sugars_100g                                100.0                    0
proteins_100g                              100.0                    0
fiber_100g                                 100.0                    0
fat_100g                                   100.0                    0
nutrition-score-fr_100g                    100.0                    0
fruits-vegetables-nuts_100g                 99.7                  181
nutrition_grade_fr                          73.7                17978

Comme prévu, il subsiste des manquants sur la variable *fruits-vegetables-nuts_100g*.

Je vérifie ce qu'il en est.

In [39]:
df_filtre.loc[df_filtre['fruits-vegetables-nuts_100g'].isna()]

code                                       product_name  \
725    0028400040112                                           Cheetos    
735    0028400090858                                               Lays   
840    0033383570051                                     Sweet Potatoes   
2733        01058963                                   Desiree potatoes   
2852        01370300                              lady baffour potatoes   
3576        03248461                 Tesco British maris piper potatoes   
4923   2000000003778                            Deluxe Potatoes Moyenne   
6363        20024451                          Pommes de terres entières   
6912        20141486                                            PopCorn   
7022        20163822                           Deutsches Kartoffelpüree   
7331        20239787                                 Pomme de terre bio   
7358        20242244                                     white potatoes   
7410        20253400                      Pommes de terre - Chair ferme   
7429        20260330                              Parmentine à rissoler   
7867        20399689                       Pommes de terres chair ferme   
8342        20545642                        La pomme de terre - frites    
9226        22014122                                    Washed Potatoes   
10269  2565633001193                                            Batatas   
10906       26061818                                  La pomme de terre   
11875  3000040048519                    Pommes de terre de consommation   
11879  3000044231658           Pommes de terre VAPEUR Campagne Première   
11905  3000124026907                                           Amandine   
11910  3000165040085                            Pomme de terre nouvelle   
18610  3168930006848                          Pop Corn Apéro, Goût Salé   
18611  3168930006855                       Pop Corn Apéro, Goût Fromage   
18612  3168930006862                      Pop Corn Apéro, Goût Barbecue   
20139  3182340919284        Pommes de terre entières cuites à la vapeur   
20476  3201922211032                    Pommes de terre de consommation   
21359  3222473000821                          Pop Corn Micro-Ondes Salé   
21463  3222473268979                           Purée de pommes de terre   
21482  3222473276141                                 Pomme de terre bio   
22314  3227291010120      Pommes de terre entières Légumes vapeur Lunor   
22315  3227291010311                         Pomme de terre en lamelles   
22316  3227291014777              Les Pommes de terre rondes à rissoler   
22317  3227291015170            Les Pommes de terre lamelles à rissoler   
23691  3245411603236  Pommes de terre de consommation bio Française Bio   
28670  3256222113715                                    Pommes de terre   
28922  3256223378373                        Pommes de terre potage four   
29923  3257980203878  Pommes de terre Four, purée, plats mijotés, po...   
32458  3263851446312                     Pommes de terre de Noirmoutier   
34003  3263859874513                           Purée de Pommes de Terre   
35342  3270160501250  Purée de pomme de terre au jambon et à l'emmental   
35344  3270160501595  Purée de pommes de terre au poulet et aux caro...   
35353  3270160504725  Purée aux 2 courgettes et pomme de terre surgelée   
35355  3270160505050  Purée de pomme de terre à la tomate et à la mo...   
35384  3270160590278  Ecrasée de Pomme de Terre aux Truffes noires (...   
35586  3270160757206                         Patates douces en morceaux   
35981  3270190028918  Pommes de terre Vapeur-Raclette Gratin-Rissolé...   
37128  3276550047711  Pommes de terre Spécial gratin - vapeur Bio Ca...   
37156  3276550322542                            pomme de tere Grenaille   
37183  3276555320314                                  Rattes du Touquet   
37189  3276555649774                            Pommes de terre Artemis   
37193  3276556201223        Pommes de ter

On a bien que des produits du type *Potatoes* ou *Salty and fatty products* pour lesquels je vais affecter la valeur 0.

In [40]:
df_filtre.loc[df_filtre['fruits-vegetables-nuts_100g'].isna(), 'fruits-vegetables-nuts_100g'] = 0

A présent, je peux finaliser les manquants de *nutrition_grade_fr* en me servant du *nutrition-score-fr_100g* en prenant garde au classement particulier concernant les boissons.

In [41]:
df_filtre['pnns_groups_1'].unique()

array(['Fruits and vegetables', 'Sugary snacks', 'Cereals and potatoes',
       'Composite foods', 'Fish meat eggs', 'Beverages', 'Fat and sauces',
       'Milk and dairy products', 'Salty snacks'], dtype=object)

In [42]:
df_filtre.loc[(df_filtre['nutrition_grade_fr'].isna())
             & (df_filtre['pnns_groups_1']=='Beverages') 
             & (df_filtre['nutrition-score-fr_100g']<=1), 'nutrition_grade_fr'] = 'b'

df_filtre.loc[(df_filtre['nutrition_grade_fr'].isna())
             & (df_filtre['pnns_groups_1']=='Beverages') 
             & (df_filtre['nutrition-score-fr_100g']<=5)
             & (df_filtre['nutrition-score-fr_100g']>1), 'nutrition_grade_fr'] = 'c'
                
df_filtre.loc[(df_filtre['nutrition_grade_fr'].isna())
             & (df_filtre['pnns_groups_1']=='Beverages') 
             & (df_filtre['nutrition-score-fr_100g']<=9)
             & (df_filtre['nutrition-score-fr_100g']>5), 'nutrition_grade_fr'] = 'd'

df_filtre.loc[(df_filtre['nutrition_grade_fr'].isna())
             & (df_filtre['pnns_groups_1']=='Beverages') 
             & (df_filtre['nutrition-score-fr_100g']>9), 'nutrition_grade_fr'] = 'e'
                
df_filtre.loc[(df_filtre['nutrition_grade_fr'].isna())
             & (df_filtre['pnns_groups_1']!='Beverages') 
             & (df_filtre['nutrition-score-fr_100g']<=-1), 'nutrition_grade_fr'] = 'a'
                
df_filtre.loc[(df_filtre['nutrition_grade_fr'].isna())
             & (df_filtre['pnns_groups_1']!='Beverages') 
             & (df_filtre['nutrition-score-fr_100g']<=2)
             & (df_filtre['nutrition-score-fr_100g']>-1), 'nutrition_grade_fr'] = 'b'

df_filtre.loc[(df_filtre['nutrition_grade_fr'].isna())
             & (df_filtre['pnns_groups_1']!='Beverages') 
             & (df_filtre['nutrition-score-fr_100g']<=10)
             & (df_filtre['nutrition-score-fr_100g']>2), 'nutrition_grade_fr'] = 'c'

df_filtre.loc[(df_filtre['nutrition_grade_fr'].isna())
             & (df_filtre['pnns_groups_1']!='Beverages') 
             & (df_filtre['nutrition-score-fr_100g']<=18)
             & (df_filtre['nutrition-score-fr_100g']>10), 'nutrition_grade_fr'] = 'd'

df_filtre.loc[(df_filtre['nutrition_grade_fr'].isna())
             & (df_filtre['pnns_groups_1']!='Beverages')
             & (df_filtre['nutrition-score-fr_100g']>18), 'nutrition_grade_fr'] = 'e'

Je vérifie que les taux de remplissage sont à 100% partout.

In [43]:
calcul_remplissage(df_filtre)

Taux de remplissage  Nombre de manquants
code                                       100.0                    0
product_name                               100.0                    0
brands                                     100.0                    0
countries_fr                               100.0                    0
pnns_groups_1                              100.0                    0
pnns_groups_2                              100.0                    0
nutrition_grade_fr                         100.0                    0
energy_100g                                100.0                    0
saturated-fat_100g                         100.0                    0
sodium_100g                                100.0                    0
sugars_100g                                100.0                    0
proteins_100g                              100.0                    0
fiber_100g                                 100.0                    0
fruits-vegetables-nuts_100g                100.0                    0
fat_100g                                   100.0                    0
nutrition-score-fr_100g                    100.0                    0

## Préparation du dataframe pour l'analyse exploratoire

In [44]:
df_filtre.to_csv("off_propre.csv", sep=',', index=False)